src/htm/algorithms/SpatialPooler.cpp

```diff
@@ -473,14 +477,40 @@ const vector<SynapseIdx> SpatialPooler::compute(const SDR &input, const bool lea
     bumpUpWeakColumns_();
     updateBoostFactors_();
     if (isUpdateRound_()) {
+      const UInt old_inhibitionRadius_ = inhibitionRadius_;
       updateInhibitionRadius_();
+      if (inhibitionRadius_ != old_inhibitionRadius_) { 
+	      calculateWrapAroundNeighbors();
+	      mapAllNeighbors();
+      }
       updateMinDutyCycles_();
     }
   }
 
   return overlaps;
 }
 
+void SpatialPooler::calculateWrapAroundNeighbors() {
+      wrapAroundNeighbors_ = 1;
+      for (UInt i = 0; i<columnDimensions_.size();i++) {
+	   UInt diam = 2*inhibitionRadius_ + 1;
+	   wrapAroundNeighbors_ *= std::min(diam, columnDimensions_[i]);
+      }
+      wrapAroundNeighbors_ -= 1;
+      return;
+}
+
+void SpatialPooler::mapAllNeighbors() {
+	neighborMap_.clear();
+	neighborMap_.reserve(wrapAroundNeighbors_ * numColumns_);
+	for(UInt column=0; column<numColumns_;column++) {
+		for(auto neighbor: WrappingNeighborhood(column, inhibitionRadius_,columnDimensions_)) { 
```

Comment:
my idea was to spend a really lot of time, mem in init, and cache this for all of `columns x inhibition_range` where cols typically 1-4k, inh radius (I'll have to see if it can be fixed, or) varies in `min(columns dim)/2`. I think we can stuff that into RAM

Reply:
Are you saying that there are more things that should be cached? What did I miss?

Reply:
How often does the inhibition radius change in real applications? I ran the "perf" tool against dynamic_hotgym with the commits in this PR and now mapAllNeighbors() (including the functions it calls) accounts for less than 1% of total run-time (plus or minus the margin of error of perf). Among the spatialPooler functions, the big villains are still updateBoostFactorsLocal_() (27% of total runtime) and inhibitColumnsLocal_() (13% of total runtime), but this has more to do with the amount of calculations they have to perform, I think.

Also, how do you figure memory requirements would scale with regards to amount of columns and disposition of columns in dimensions?

Reply:
> How often does the inhibition radius change in real applications?  

every `isUpdateRound_` which defaults to 500, so 1/500th of the time. We've almost never changed that default. 

> now mapAllNeighbors() (including the functions it calls) accounts for less than 1% of total run-time 

ok, I agree this is good enough and not worth complicating things. Leave as is. 

>  the big villains are still updateBoostFactorsLocal_() (27% of total runtime) 

this is where the cached Neighborhood should help, all the function does is just looping through the neighborhood and increasing counter to get `numNeighbors`, I'd like to get it in O(1) 

> there can be easily a single map that goes from minInhibitionRadius_ to maxInhibitionRadius_, so RAM isn't really a concern, but without wrapAround this becomes more difficult 

I'd suggest not to worry about `wrapAround=False` case (unless you can do it easily too). If these patches work well, I'll suggest keeping only the `WrappingNeighborhood` (unless there are proven use-cases for the non-wrapping).